src/db/connection.rs

```diff
@@ -99,17 +122,89 @@ impl<'a, 'r> rocket::request::FromRequest<'a, 'r> for RequestContext {
 pub struct DatabaseContext<'a> {
     // Eventually make this private
     pub conn: &'a PgConnection,
-    action: ActionType,
-    modifier: ActionModifier,
+    pub resource: &'a str,
+    pub auth: Authorization,
+    pub action: ActionType,
+    pub modifier: ActionModifier,
 }
 
 impl<'a> DatabaseContext<'a> {
+    fn try_get_permissions(
+        conn: &'a PgConnection,
+        resource: &str,
+        token: Option<Token>,
+        action: ActionType,
+        modifier: ActionModifier,
+    ) -> Result<Authorization> {
+        use crate::resolvers::permission;
+        use crate::resolvers::user;
+        if let Some(token) = token {
+            let user = user::get_by_email(&token.uid, conn)?;
+            let permissions = permission::get_permission(conn, &user, resource, action, modifier)?;
+            if permissions.is_empty() {
+                Ok(Authorization::UnAuthorized { user })
+            } else {
+                Ok(Authorization::Authorized { user, permissions })
+            }
+        } else {
+            Ok(Authorization::Anon)
+        }
+    }
     /// Create a DatabaseContext Struct from a connection and modifier information
-    fn from(conn: &'a PgConnection, action: ActionType, modifier: ActionModifier) -> Self {
-        Self {
+    fn from(
+        conn: &'a PgConnection,
+        resource: &'a str,
+        token: Option<Token>,
+        action: ActionType,
+        modifier: ActionModifier,
+    ) -> Result<Self> {
+        let auth = Self::try_get_permissions(conn, resource, token, action, modifier)?;
+        Ok(Self {
             conn,
+            auth,
+            resource,
             action,
             modifier,
+        })
+    }
+
+    pub fn get_user(&self) -> Result<&User> {
+        match &self.auth {
+            &Authorization::Anon => Err(TokenError::Expired)?,
```

Comment:
Let's not give an expired error as that's not really what happened. How about we create a new error variant?

Reply:
Fixed!